File: 37-sudoku-solver/37-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {

		solve(board, 0, 0);

	}
    
    private boolean solve(char[][] board, int i, int j) {

		if (i == board.length) {

			
			return true;
		}

		int ni = 0;
		int nj = 0;

		// if j reaches end, initialize it with next i=i+1,j=0
		if (j == board.length - 1) {

			nj = 0;
			ni = i + 1;
		} else {

			ni = i;
			nj = j + 1;
		}

		// if content on board at i and j is not null, advance to next cell
		if (board[i][j] == '.') {

			for (char po = '1'; po <= '9'; po++) {

				// if success

				if (isValid(board, i, j, po)) {

					// set board value with that num
					board[i][j] = po;
					if (solve(board, ni, nj)) {
						return true;
					}

					board[i][j] = '.';

				}
			}

		}else {
			return solve(board, ni, nj);
		}
		
		return false;
	}

	private static boolean isValid(char[][] board, int x, int y, char val) {

		// check in row
		int n = board.length;
		for (int i = 0; i < n; i++) {

			if (board[i][y] == val) {
				return false;
			}
		}

		// check in col
		for (int i = 0; i < n; i++) {

			if (board[x][i] == val) {
				return false;
			}
		}

		// check in submatrix

		int smi = (x / 3) * 3;
		int smj = (y / 3) * 3;

		for (int i = 0; i < 3; i++) {

			for (int j = 0; j < 3; j++) {

				if (board[smi + i][smj + j] == val) {

					return false;
				}

			}
		}
        return true;

    }	
}

File: 37-sudoku-solver/37-sudoku-solver.cpp
class Solution {
public:
    bool isValid(char val, vector<vector<char>> &board,int row,int col){
        // row check
        for(int i=0; i<board[0].size(); ++i){
            if(board[row][i] == val){
                return false;
            }
        }
        //col check
        for(int i=0; i<board.size(); ++i){
            if(board[i][col] == val){
                return false;
            }
        }
        // 3*3 matrix check
        int sub_mat_i = (row / 3) * 3;
        int sub_mat_j = (col / 3) * 3;
        for(int i=0; i<3; ++i)
        {
            for(int j=0; j<3; ++j)
            {
                if(board[sub_mat_i + i][sub_mat_j + j] == val)
                {
                    return false;
                }
            }
        }
        return true;
    }
    
    bool sudokuSolver(vector<vector<char>> &board,int row,int col){
        if(row == board.size()) return true;
        
        int next_row,next_col ;
        
        if(col == board[0].size()-1)
        {
            next_col = 0;
            next_row = row+1;
        }
        else
        {
            next_col = col + 1;
            next_row = row;
        }
        
        if(board[row][col] != '.')
        {
            return sudokuSolver(board,next_row,next_col);
        }
        else{
            for(char po='1'; po<='9'; ++po)
            {
                if(isValid(po,board,row,col))
                {
                    board[row][col] = po;
                    if(sudokuSolver(board,next_row,next_col))
                    {
                        return true;
                    }
    
                    board[row][col] = '.';
                }
            } 
        }
        return false;
    }
    void solveSudoku(vector<vector<char>>& board)
    {
        
        int row = 0; 
        int col = 0;
        
        sudokuSolver(board,row,col);
        
    }
};